drivers/network/netio/netio.c

```diff
@@ -0,0 +1,896 @@
+            s->LocalAddressFile = NULL;
+        }
+        if (s->ConnectionHandle != INVALID_HANDLE_VALUE)
+        {
+            ZwClose(s->ConnectionHandle);
+            s->ConnectionHandle = INVALID_HANDLE_VALUE;
+            s->ConnectionFile = NULL;
+        }
+        /* TODO: Free TdiName and others ... */
+        ExFreePoolWithTag(s, 'SOCK');
+    }
+}
+
+        /* This just sets the status to pending. It is been called
+         * by the IoCallDriver() function via the MajorFunction
+         * dispatch table in the DriverObject. We need this because
+         * we want to use IoCallDriver to consume one stack location
+         * of the irp. If we do not do so the completion routine of
+         * the upper driver never gets called.
+         */
+
+static NTSTATUS NTAPI
+DummyHandler(PDEVICE_OBJECT Device, PIRP Irp)
+{
+    Irp->IoStatus.Status = STATUS_PENDING;
+
+    return STATUS_PENDING;
+}
+
+struct DummyDeviceExtension
+{
+    int x;
+};
+
+PDEVICE_OBJECT DummyDeviceObject;
+
+static NTSTATUS
+DummyCallDriver(PIRP Irp)
+{
+    return IoCallDriver(DummyDeviceObject, Irp);
+}
+
+NTSTATUS NTAPI
+DriverEntry(PDRIVER_OBJECT DriverObject, PUNICODE_STRING RegistryPath)
+{
+    int i;
+    NTSTATUS status;
+    UNICODE_STRING nameUnicode;
+
+    DbgPrint("netio.sys DriverEntry compiled " __DATE__ " " __TIME__ "\n");
+
+    for(i = 0; i <= IRP_MJ_MAXIMUM_FUNCTION; i++)
+        DriverObject->MajorFunction[i] = DummyHandler;
+
+    RtlInitUnicodeString(&nameUnicode, L"\\Device\\NetioSysDummy");
+
+    status = IoCreateDevice(DriverObject, sizeof(struct DummyDeviceExtension),
+                            &nameUnicode, FILE_DEVICE_UNKNOWN,
+                            FILE_DEVICE_SECURE_OPEN, FALSE, &DummyDeviceObject);
+
+    if (!NT_SUCCESS(status))
+    {
+        DbgPrint("Can't create root, err=%x\n", status);
+        return status;
+    }
+    return STATUS_SUCCESS;
+}
+
+static NTSTATUS NTAPI
+NetioComplete(PDEVICE_OBJECT DeviceObject, PIRP Irp, PVOID Context)
+{
+    struct NetioContext *c = (struct NetioContext *)Context;
+    PIRP UserIrp = c->UserIrp;
+
+    UserIrp->IoStatus.Status = Irp->IoStatus.Status;
+    UserIrp->IoStatus.Information = Irp->IoStatus.Information;
+
+    RemoveEntryList(&UserIrp->Tail.Overlay.ListEntry);
+    IoCompleteRequest(UserIrp, IO_NETWORK_INCREMENT);
+
+    SocketPut(c->socket);
+    ExFreePoolWithTag(c, 'NEIO');
+
+    return STATUS_SUCCESS;
+}
+
+static NTSTATUS WSKAPI
+WskControlSocket(_In_ PWSK_SOCKET Socket,
+                 _In_ WSK_CONTROL_SOCKET_TYPE RequestType,
+                 _In_ ULONG ControlCode,
+                 _In_ ULONG Level,
+                 _In_ SIZE_T InputSize,
+                 _In_reads_bytes_opt_(InputSize) PVOID InputBuffer,
+                 _In_ SIZE_T OutputSize,
+                 _Out_writes_bytes_opt_(OutputSize) PVOID OutputBuffer,
+                 _Out_opt_ SIZE_T * OutputSizeReturned, _Inout_opt_ PIRP Irp)
```

Comment:
```suggestion
static NTSTATUS WSKAPI
WskControlSocket(
    _In_ PWSK_SOCKET Socket,
    _In_ WSK_CONTROL_SOCKET_TYPE RequestType,
    _In_ ULONG ControlCode,
    _In_ ULONG Level,
    _In_ SIZE_T InputSize,
    _In_reads_bytes_opt_(InputSize) PVOID InputBuffer,
    _In_ SIZE_T OutputSize,
    _Out_writes_bytes_opt_(OutputSize) PVOID OutputBuffer,
    _Out_opt_ SIZE_T * OutputSizeReturned, _Inout_opt_ PIRP Irp)
```
avoid depending on the func names length for indent.